### PR TITLE
treewide: relax numba version

### PR DESCRIPTION
**Pull request by natsukium**
## Description of changes

This is a follow up PR for #246727.

The version of numba is currently not set. As a result, packages that require numba with the lower limit failed to build because they could not find the appropriate numba.
This is a temporary fix until numba is updated to a newer version.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by natsukium** [comment]
Result of `nixpkgs-review pr 248314` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages marked as broken and skipped:</summary>
  <ul>
    <li>piper-train</li>
    <li>piper-train.dist</li>
    <li>python310Packages.qiskit</li>
    <li>python310Packages.qiskit-machine-learning</li>
    <li>python310Packages.qiskit-machine-learning.dist</li>
    <li>python310Packages.qiskit.dist</li>
    <li>python310Packages.scikit-tda</li>
    <li>python310Packages.scikit-tda.dist</li>
  </ul>
</details>
<details>
  <summary>10 packages failed to build:</summary>
  <ul>
    <li>pentestgpt</li>
    <li>pentestgpt.dist</li>
    <li>python310Packages.coffea</li>
    <li>python310Packages.coffea.dist</li>
    <li>python310Packages.dask-glm</li>
    <li>python310Packages.dask-glm.dist</li>
    <li>python310Packages.dask-ml</li>
    <li>python310Packages.dask-ml.dist</li>
    <li>python310Packages.langchain</li>
    <li>python310Packages.langchain.dist</li>
  </ul>
</details>
<details>
  <summary>31 packages built:</summary>
  <ul>
    <li>pianotrans</li>
    <li>pianotrans.dist</li>
    <li>pitivi</li>
    <li>python310Packages.awkward</li>
    <li>python310Packages.awkward.dist</li>
    <li>python310Packages.clifford</li>
    <li>python310Packages.clifford.dist</li>
    <li>python310Packages.dask-awkward</li>
    <li>python310Packages.dask-awkward.dist</li>
    <li>python310Packages.librosa</li>
    <li>python310Packages.librosa.dist</li>
    <li>python310Packages.mplhep</li>
    <li>python310Packages.mplhep.dist</li>
    <li>python310Packages.piano-transcription-inference</li>
    <li>python310Packages.piano-transcription-inference.dist</li>
    <li>python310Packages.pynndescent</li>
    <li>python310Packages.pynndescent.dist</li>
    <li>python310Packages.resampy</li>
    <li>python310Packages.resampy.dist</li>
    <li>python310Packages.sparse</li>
    <li>python310Packages.sparse.dist</li>
    <li>python310Packages.torchlibrosa</li>
    <li>python310Packages.torchlibrosa.dist</li>
    <li>python310Packages.umap-learn</li>
    <li>python310Packages.umap-learn.dist</li>
    <li>python310Packages.uproot</li>
    <li>python310Packages.uproot.dist</li>
    <li>python310Packages.vector</li>
    <li>python310Packages.vector.dist</li>
    <li>tts</li>
    <li>tts.dist</li>
  </ul>
</details>

**Comment by doronbehar** [comment]
Is it possible to use something like this to spare some of the patches maybe?

```diff
diff --git i/pkgs/development/python-modules/numba/default.nix w/pkgs/development/python-modules/numba/default.nix
index da0f72a43375..61dd95b582b6 100644
--- i/pkgs/development/python-modules/numba/default.nix
+++ w/pkgs/development/python-modules/numba/default.nix
@@ -52,6 +52,10 @@ in buildPythonPackage rec {
     hash = "sha256-Wm1sV4uS/Xkz1BkT2xNmwgBZS0X8YziC6jlbfolXGB8=";
   };
   env.NIX_CFLAGS_COMPILE = lib.optionalString stdenv.isDarwin "-I${lib.getDev libcxx}/include/c++/v1";
+  preConfigure = ''
+    substituteInPlace numba/_version.py \
+      --replace 'git_refnames = ""' 'git_refnames = "0.58.0-rc"'
+  '';
 
   nativeBuildInputs = [
     numpy
```

If not, then this PR looks mostly good, (although I would rephrase some of the comments).

**Comment by natsukium** [comment]
This patch looks fine.
I think I can fix most of the changes in this PR, and I will resubmit it.
Thanks for your suggestion!

**Comment by natsukium** [comment]
successor #248316
